File: android/app/src/main/java/com/noop/ui/BackupSync.kt
package com.noop.ui

import android.content.Context
import android.net.Uri
import android.provider.DocumentsContract
import androidx.work.CoroutineWorker
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.WorkerParameters
import com.noop.data.DataBackup
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.TimeUnit

/**
 * Backup & Sync (Phase 1 — folder destination). Writes the full `.noopbak` snapshot (the existing
 * [DataBackup] format) into a user-chosen folder (a SAF tree), on demand and on a daily schedule.
 * Point that folder at your Google Drive / iCloud / Dropbox desktop-sync client and you get automatic
 * off-device backup with no in-app cloud account, no OAuth, no secrets — NOOP only writes a local file;
 * your sync client does the upload.
 *
 * DESIGN: a future Google-Drive destination slots in behind [Destination] (this is the FolderDestination
 * leg). Snapshots are timestamped + immutable; "restore" REPLACES the local DB (it's a whole-DB snapshot,
 * not a record merge — newest-wins), matching DataBackup. The pure filename helpers are unit-tested.
 */
object BackupSync {

    /** The two backup destinations Phase 1 is built to support (Drive lands in Phase 2). */
    enum class Destination { FOLDER /* , GOOGLE_DRIVE */ }

    private const val PREFIX = "noop-backup-"
    private const val SUFFIX = ".noopbak"
    const val MIME = "application/octet-stream"

    // ── Pure helpers (unit-tested) ──────────────────────────────────────────

    private fun fmt() = java.text.SimpleDateFormat("yyyyMMdd-HHmmss", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    /** Canonical snapshot filename for an instant: `noop-backup-YYYYMMDD-HHMMSS.noopbak` (UTC). */
    fun snapshotName(epochMs: Long): String = PREFIX + fmt().format(java.util.Date(epochMs)) + SUFFIX

    /** The UTC instant encoded in a snapshot filename, or null if [name] is not one of ours. */
    fun snapshotTimeMs(name: String): Long? {
        if (!name.startsWith(PREFIX) || !name.endsWith(SUFFIX)) return null
        val stamp = name.substring(PREFIX.length, name.length - SUFFIX.length)
        return runCatching { fmt().parse(stamp)?.time }.getOrNull()
    }

    fun isSnapshot(name: String): Boolean = snapshotTimeMs(name) != null

    /** Newest snapshot by encoded time among [names] (non-snapshots ignored), or null if none. */
    fun latestSnapshot(names: List<String>): String? =
        names.filter(::isSnapshot).maxByOrNull { snapshotTimeMs(it)!! }

    /** Snapshots to DELETE to keep only the [keep] newest (oldest-first). Empty when within budget. */
    fun snapshotsToPrune(names: List<String>, keep: Int): List<String> {
        val snaps = names.filter(::isSnapshot).sortedByDescending { snapshotTimeMs(it)!! }
        return if (snaps.size <= keep) emptyList() else snaps.drop(keep)
    }

    // ── Folder destination I/O (SAF tree via DocumentsContract — no extra dep) ───

    /** Create + write one snapshot into the chosen [treeUri]; returns the new file Uri, or null on failure. */
    fun writeSnapshot(context: Context, treeUri: Uri, nowMs: Long = System.currentTimeMillis()): Uri? {
        val resolver = context.contentResolver
        val parentDoc = DocumentsContract.buildDocumentUriUsingTree(
            treeUri, DocumentsContract.getTreeDocumentId(treeUri),
        )
        val fileUri = runCatching {
            DocumentsContract.createDocument(resolver, parentDoc, MIME, snapshotName(nowMs))
        }.getOrNull() ?: return null
        return runCatching { DataBackup.exportTo(context, fileUri); fileUri }.getOrNull()
    }

    /** Run one backup into the persisted folder, prune to [BackupSyncPrefs.keepCount], stamp last-backup. */
    fun backupNow(context: Context): Boolean {
        val treeUri = BackupSyncPrefs.treeUri(context) ?: return false
        val written = writeSnapshot(context, treeUri) ?: return false
        BackupSyncPrefs.setLastBackupMs(context, System.currentTimeMillis())
        prune(context, treeUri)
        return written != Uri.EMPTY
    }

    /** Best-effort retention: delete snapshots beyond keepCount. Listing failures are ignored. */
    private fun prune(context: Context, treeUri: Uri) {
        val keep = BackupSyncPrefs.keepCount(context)
        val children = runCatching { listSnapshotDocs(context, treeUri) }.getOrDefault(emptyList())
        val toDelete = snapshotsToPrune(children.map { it.first }, keep).toSet()
        for ((name, docUri) in children) {
            if (name in toDelete) runCatching { DocumentsContract.deleteDocument(context.contentResolver, docUri) }
        }
    }

    /** (name, documentUri) for every snapshot in the tree, newest-first. */
    fun listSnapshotDocs(context: Context, treeUri: Uri): List<Pair<String, Uri>> {
        val childrenUri = DocumentsContract.buildChildDocumentsUriUsingTree(
            treeUri, DocumentsContract.getTreeDocumentId(treeUri),
        )
        val out = ArrayList<Pair<String, Uri>>()
        context.contentResolver.query(
            childrenUri,
            arrayOf(DocumentsContract.Document.COLUMN_DOCUMENT_ID, DocumentsContract.Document.COLUMN_DISPLAY_NAME),
            null, null, null,
        )?.use { c ->
            while (c.moveToNext()) {
                val id = c.getString(0); val name = c.getString(1) ?: continue
                if (isSnapshot(name)) out.add(name to DocumentsContract.buildDocumentUriUsingTree(treeUri, id))
            }
        }
        return out.sortedByDescending { snapshotTimeMs(it.first)!! }
    }

    // ── Scheduling (WorkManager — survives reboot/app-kill, mirrors DebugExportScheduler) ──

    private const val WORK = "noop-backup-sync"

    fun reschedule(context: Context) {
        val wm = WorkManager.getInstance(context.applicationContext)
        if (!BackupSyncPrefs.autoEnabled(context) || BackupSyncPrefs.treeUri(context) == null) {
            wm.cancelUniqueWork(WORK); return
        }
        val req = PeriodicWorkRequestBuilder<BackupSyncWorker>(1, TimeUnit.DAYS).build()
        wm.enqueueUniquePeriodicWork(WORK, ExistingPeriodicWorkPolicy.KEEP, req)
    }
}

class BackupSyncWorker(appContext: Context, params: WorkerParameters) :
    CoroutineWorker(appContext, params) {
    override suspend fun doWork(): Result {
        if (!BackupSyncPrefs.autoEnabled(applicationContext)) return Result.success()
        return if (BackupSync.backupNow(applicationContext)) Result.success() else Result.retry()
    }
}

/** Small SharedPreferences-backed store for the folder destination + schedule state. */
object BackupSyncPrefs {
    private const val FILE = "backup_sync"
    private fun p(c: Context) = c.applicationContext.getSharedPreferences(FILE, Context.MODE_PRIVATE)

    fun treeUri(c: Context): Uri? = p(c).getString("tree_uri", null)?.let(Uri::parse)
    fun setTreeUri(c: Context, uri: Uri?) = p(c).edit().apply {
        if (uri == null) remove("tree_uri") else putString("tree_uri", uri.toString())
    }.apply()

    fun autoEnabled(c: Context): Boolean = p(c).getBoolean("auto", false)
    fun setAutoEnabled(c: Context, on: Boolean) = p(c).edit().putBoolean("auto", on).apply()

    fun lastBackupMs(c: Context): Long = p(c).getLong("last_ms", 0L)
    fun setLastBackupMs(c: Context, ms: Long) = p(c).edit().putLong("last_ms", ms).apply()

    fun keepCount(c: Context): Int = p(c).getInt("keep", 10)
    fun setKeepCount(c: Context, n: Int) = p(c).edit().putInt("keep", n.coerceIn(1, 100)).apply()
}

File: Strand/Data/BackupSync.swift
import Foundation
#if os(macOS)
import AppKit
#else
import UIKit
import UniformTypeIdentifiers
#endif

/// Backup & Sync (Phase 1 — folder destination), the Apple mirror of the Android `BackupSync`.
/// Writes the full `.noopbak` snapshot (existing `DataBackup` format) into a user-chosen folder, on
/// demand and as an on-launch catch-up. Point that folder at a Google Drive / iCloud / Dropbox client
/// for automatic off-device backup with no in-app cloud account — NOOP only writes a local file.
///
/// Built behind `Destination` so a Phase-2 Google Drive backend slots in behind the same UI/logic.
/// The pure filename/selection helpers are unit-tested byte-for-byte against the Android twin.
enum BackupSync {

    enum Destination { case folder /* , googleDrive */ }

    static let prefix = "noop-backup-"
    static let suffix = ".noopbak"

    // ── Pure helpers (unit-tested; mirror Android BackupSync) ────────────────

    private static let stampFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyyMMdd-HHmmss"
        f.isLenient = false
        return f
    }()

    /// Canonical snapshot filename for an instant (ms since epoch): `noop-backup-YYYYMMDD-HHMMSS.noopbak` (UTC).
    static func snapshotName(_ epochMs: Int) -> String {
        prefix + stampFormatter.string(from: Date(timeIntervalSince1970: Double(epochMs) / 1000.0)) + suffix
    }

    /// The UTC instant (ms) encoded in a snapshot filename, or nil if not one of ours.
    static func snapshotTimeMs(_ name: String) -> Int? {
        guard name.hasPrefix(prefix), name.hasSuffix(suffix) else { return nil }
        let stamp = String(name.dropFirst(prefix.count).dropLast(suffix.count))
        guard let d = stampFormatter.date(from: stamp) else { return nil }
        return Int((d.timeIntervalSince1970 * 1000.0).rounded())
    }

    static func isSnapshot(_ name: String) -> Bool { snapshotTimeMs(name) != nil }

    /// Newest snapshot by encoded time (non-snapshots ignored), or nil.
    static func latestSnapshot(_ names: [String]) -> String? {
        names.filter(isSnapshot).max { (snapshotTimeMs($0) ?? 0) < (snapshotTimeMs($1) ?? 0) }
    }

    /// Snapshots to DELETE to keep only the `keep` newest (oldest-first). Empty when within budget.
    static func snapshotsToPrune(_ names: [String], keep: Int) -> [String] {
        let snaps = names.filter(isSnapshot).sorted { (snapshotTimeMs($0) ?? 0) > (snapshotTimeMs($1) ?? 0) }
        return snaps.count <= keep ? [] : Array(snaps.dropFirst(keep))
    }
}

/// The folder destination: security-scoped bookmark of a user-chosen folder + write/restore + an
/// on-launch catch-up (Apple sandbox/background limits make a daily catch-up the pragmatic schedule —
/// macOS already relies on foreground for scheduled exports; iOS backs up on next foreground).
@MainActor
enum FolderBackup {
    private static let bookmarkKey = "backupSync.folderBookmark"
    private static let autoKey = "backupSync.auto"
    private static let lastKey = "backupSync.lastMs"
    static let keepCount = 10
    private static let dayMs = 24 * 60 * 60 * 1000

    static var autoEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: autoKey) }
        set { UserDefaults.standard.set(newValue, forKey: autoKey) }
    }
    static var lastBackupMs: Int { UserDefaults.standard.integer(forKey: lastKey) }
    static var hasFolder: Bool { UserDefaults.standard.data(forKey: bookmarkKey) != nil }

    static func folderLabel() -> String? { resolveFolder()?.lastPathComponent }

    private static func bookmarkOptions() -> URL.BookmarkCreationOptions {
        #if os(macOS)
        return [.withSecurityScope]
        #else
        return []
        #endif
    }

    private static func resolveFolder() -> URL? {
        guard let data = UserDefaults.standard.data(forKey: bookmarkKey) else { return nil }
        var stale = false
        #if os(macOS)
        let opts: URL.BookmarkResolutionOptions = [.withSecurityScope]
        #else
        let opts: URL.BookmarkResolutionOptions = []
        #endif
        return try? URL(resolvingBookmarkData: data, options: opts, relativeTo: nil, bookmarkDataIsStale: &stale)
    }

    static func saveFolder(_ url: URL) {
        let scoped = url.startAccessingSecurityScopedResource()
        defer { if scoped { url.stopAccessingSecurityScopedResource() } }
        if let data = try? url.bookmarkData(options: bookmarkOptions(), includingResourceValuesForKeys: nil, relativeTo: nil) {
            UserDefaults.standard.set(data, forKey: bookmarkKey)
        }
    }

    /// Write one snapshot into the bookmarked folder; returns true on success.
    static func backupNow(checkpoint: @escaping () async -> Bool) async -> Bool {
        guard let folder = resolveFolder() else { return false }
        let scoped = folder.startAccessingSecurityScopedResource()
        defer { if scoped { folder.stopAccessingSecurityScopedResource() } }
        let nowMs = Int(Date().timeIntervalSince1970 * 1000.0)
        let dest = folder.appendingPathComponent(BackupSync.snapshotName(nowMs))
        guard case .exported = await DataBackup.writeBackup(checkpoint: checkpoint, to: dest) else { return false }
        UserDefaults.standard.set(nowMs, forKey: lastKey)
        prune(in: folder)
        return true
    }

    private static func prune(in folder: URL) {
        let names = (try? FileManager.default.contentsOfDirectory(atPath: folder.path)) ?? []
        let toDelete = Set(BackupSync.snapshotsToPrune(names, keep: keepCount))
        for name in names where toDelete.contains(name) {
            try? FileManager.default.removeItem(at: folder.appendingPathComponent(name))
        }
    }

    /// On-launch catch-up: if auto is on, a folder is set, and it's been ≥ a day, back up.
    static func catchUpIfDue(checkpoint: @escaping () async -> Bool) async {
        guard autoEnabled, hasFolder else { return }
        if Int(Date().timeIntervalSince1970 * 1000.0) - lastBackupMs >= dayMs {
            _ = await backupNow(checkpoint: checkpoint)
        }
    }

    // ── Folder picker ────────────────────────────────────────────────────────
    #if os(macOS)
    /// Present an NSOpenPanel to choose a directory; persists the bookmark. Returns the chosen URL.
    static func pickFolder() -> URL? {
        let panel = NSOpenPanel()
        panel.canChooseDirectories = true
        panel.canChooseFiles = false
        panel.allowsMultipleSelection = false
        panel.prompt = "Choose"
        panel.message = "Choose a folder for NOOP backups (e.g. a Google Drive / iCloud folder)."
        guard panel.runModal() == .OK, let url = panel.url else { return nil }
        saveFolder(url)
        return url
    }
    #else
    /// Present a folder picker (UIDocumentPicker) and persist the bookmark.
    static func pickFolder() async -> URL? {
        let url = await DocumentPicker.pickFolder()
        if let url { saveFolder(url) }
        return url
    }
    #endif
}
